keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,558 @@
+The proposed integration here is only covering first class, non provider-specific QoS parameters(capacity, iops, throughput). It does not cover changing provider-specific parameters and first class QoS parameters all at once. The mitigation proposed is that providers can create their own CRD to address provider-specific change logic.
+
+## Design Details
+
+As part of this proposal, we are mainly proposing the following changes:
+
+1. Add new resources ResourceProvisionedIOPS and ResourceProvisionedThroughput in the existing PVC API called - pvc.Status.AllocatedResources for the purpose of tracking ProvisionedIOPS and ProvisionedThroughput. By default only the api-server or resize-controller can set this field.
+
+2. Relax API validation on PVC update so that updating ResourceProvisionedIOPS and ResourceProvisionedThroughput pvc.Spec.Resoures is allowed, as long as requested size is within the range specified by the cluster admin.
+
+3. Add new statuses in PVC API called - pvc.Status.AllocatedResourceStatus for the purpose of tracking status of ProvisionedIOPS and ProvisionedThroughput change. Existing fields can be found [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L520-L534). The new statuses will be added are:
+
+```go
+type ClaimResourceStatus string
+const (
+    PersistentVolumeClaimControllerUpdateIopsProgress ClaimResourceStatus = "ControllerUpdateIopsInProgress"
+    PersistentVolumeClaimControllerUpdateIopsPending ClaimResourceStatus = "ControllerUpdateIopsPending"
+    PersistentVolumeClaimControllerUpdateIopsFailed ClaimResourceStatus = "ControllerUpdateIopsFailed"
+    PersistentVolumeClaimControllerUpdateThroughputInProgress ClaimResourceStatus = "ControllerUpdateThroughputInProgress"
+    PersistentVolumeClaimControllerUpdateThroughputPending ClaimResourceStatus = "ControllerUpdateThroughputPending"
+    PersistentVolumeClaimControllerUpdateThroughputFailed ClaimResourceStatus = "ControllerUpdateThroughputFailed"
+)
+// PersistentVolumeClaimStatus represents the status of PV claim
+type PersistentVolumeClaimStatus struct {
+    ...
+    // AllocatedResourceStatus stores a map of resource that is being expanded
+    // and possible status that the resource exists in.
+    // Some examples may be:
+    //  pvc.status.allocatedResourceStatus["storage"] = "ControllerResizeInProgress"
+    //  pvc.status.allocatedResourceStatus["iops"] = "ControllerUpdateIopsInProgress/ControllerUpdateIopsPending/ControllerUpdateIopsFailed"
+    //  pvc.status.allocatedResourceStatus["throughput"] = "ControllerUpdateThroughputInProgress/ControllerUpdateThroughputPending/ControllerUpdateThroughputFailed"
+    AllocatedResourceStatus map[ResourceName]ClaimResourceStatus
+}
+```
+
+
+
+4. Add new CSI API ControllerUpdateIOProvisioning, when there is a change of iops/throughput in PVC, external-resizer triggers a ControllerUpdateIOProvisioning operations against a CSI endpoint.
```

Comment:
What happens if a user modifies storage capacity and QoS parameters at the same time? Will volume expansion and QoS changes trigger 2 different RPC calls against the driver (ControllerExpandVolume and ControllerUpdateIOProvisioning) or will the RPC calls be coalesced into one? For various SPs (e.g., AWS, Azure), these RPC calls translate to the same upstream API call.